### PR TITLE
Interactivity features to update the Node and Client info panels

### LINKED ISSUE #10: Implement Node info panel
Update the node info panel when a node is clicked in the graph

### LINKED ISSUE #15: Implement Client/User Journey info panel
Use graph/dropdown panel to select a Client, then display all relevant User Journey information in a table.

### PR DESCRIPTION
**Pull request by riverchen99**
Implements interactivity to update Node info panel and Client info panel. Apologies for another PR with so many changes -- I changed some linting settings and it did a bunch of reformatting automatically. Next time I make linting changes, I'll make them in a separate branch to reduce the PR size.

Also added some unit tests for functions that weren't tested previously. A lot of the lines in the PR are just static data for testing.

closes #10 
closes #15

### CODE REVIEW

File: ujt/constants.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+from graph_structures_pb2 import NodeType, Status
```
Comment (riverchen99) [inline ujt/constants.py:1]: New file to handle constants (mostly styling options)

File: ujt/converters.py [hunk 2]
```diff
@@ -75,3 +91,86 @@ def cytoscape_elements_from_clients(client_name_message_map: Dict[str, Client]):
                 })
 
     return node_elements + edge_elements
+
+
+def datatable_from_nodes(nodes, use_relative_names, table_id):
```
Comment (riverchen99) [inline ujt/converters.py:96]: Here are new functions to generate the Dash datatables that display information about nodes.

File: ujt/utils.py [hunk 3]
```diff
@@ -68,3 +69,15 @@ def read_proto_from_file(file_name: str, proto_type: Type[Message]) -> Message:
               "r") as f:
         proto_text: str = f.read()
         return text_format.Parse(proto_text, proto_type())
+
+
+def is_client_cytoscape_node(tap_node):
```
Comment (riverchen99) [inline ujt/utils.py:74]: I have a few more utility functions now -- we could split this file into proto_utils and utils if necessary? @KenArthur

File: ujt/app.py [hunk 4]
```diff
@@ -116,109 +150,191 @@ def generate_graph_elements(node_name_message_map: Dict[str, Node],
...
     """
 
-    compute_status.compute_node_statuses(node_name_message_map,
-                                         client_name_message_map)
+    return (
+        converters.cytoscape_elements_from_nodes(node_name_message_map) +
+        converters.cytoscape_elements_from_clients(client_name_message_map))
+
+
+def generate_dropdown_options(client_name_message_map: Dict[str, Client]):
```
Comment (riverchen99) [inline ujt/app.py:158]: this function is moved to converters.py in a later commit

File: ujt/app.py [hunk 5]
```diff
@@ -116,109 +150,191 @@ def generate_graph_elements(node_name_message_map: Dict[str, Node],
...
+    return cytoscape_graph_elements, client_dropdown_options
+
+
+@app.callback(
+    Output("node-info-panel",
+           "children"),
+    [Input("cytoscape-graph",
+           "tapNode")],
+)
+def generate_node_info_panel(tap_node):
```
Comment (riverchen99) [inline ujt/app.py:197]: Here's the new code to generate the node info panel including up to 3 Datatables holding data describing the Node's SLIs, Dependencies, and Children

File: ujt/app.py [hunk 6]
```diff
@@ -116,109 +150,191 @@ def generate_graph_elements(node_name_message_map: Dict[str, Node],
...
+    return converters.datatable_from_client(client, "datatable-client")
+
+
+@app.callback(
+    Output("client-dropdown",
+           "value"),
+    [Input("cytoscape-graph",
+           "tapNode")],
+)
+def update_client_dropdown_value(tap_node):
```
Comment (riverchen99) [inline ujt/app.py:283]: when the user clicks a client Node in the graph, update the value in the dropdown menu

Comment (KenArthur) [reply ujt/app.py:283]: If it's worth having a comment explaining what the function does in the review, it's worth having a comment in the code on the method itself.

File: ujt/compute_status.py [hunk 7]
```diff
@@ -26,18 +29,24 @@ def compute_node_statuses(node_name_message_map: Dict[str, Node],
...
     for client in client_name_message_map.values():
         for user_journey in client.user_journeys:
+            status_count_map: Dict["StatusValue", int] = defaultdict(int)
             for dependency in user_journey.dependencies:
-                compute_single_node_status(node_name_message_map,
-                                           dependency.target_name)
+                status_count_map[compute_single_node_status(
+                    node_name_message_map,
+                    dependency.target_name)] += 1
+            user_journey.status = compute_status_from_count_map(
```
[the remark this answers is not in the document]
Comment (KenArthur) [reply ujt/compute_status.py:37]: I think this makes sense. Realistically your graphs are always going to be relatively small and traversal probably won't be hugely expensive, so I would say if it's hard to follow then split it up and prioritize code health over speed. In this case I think it's easy enough to follow that it's not an issue.

**Review by KenArthur — APPROVED** [review]

File: ujt/app.py [hunk 8]
```diff
@@ -71,22 +76,35 @@ def read_local_data() -> Tuple[Dict[str, Node], Dict[str, Client]]:
...
-
-    client_name_message_map: Dict[str, Client] = {
-        name: cast(
-            Client,
-            utils.read_proto_from_file(
-                utils.named_proto_file_name(name, Client), Client))
-        for name in client_names
-    }
+    node_name_message_map: Dict[str,
+                                Node] = {
```
Comment (pwat2) [inline ujt/app.py:80]: is this really what the linter did? one argument on each line seems excessive

**Review by pwat2 — APPROVED** [review]

File: ujt/app.py (hunk 8 above)
Comment (riverchen99) [reply ujt/app.py:80]: I messed with some settings... I think if a line breaks the length limit it'll break on every open paren and comma? I'm still not happy with the formatting settings so I plan on adding [black](https://github.com/psf/black) eventually.
